cogs/permissions.py

```diff
@@ -0,0 +1,439 @@
+        new_permission_config = \
+            PermissionConfig.get_config(is_enabled=True, whitelisted_channel_ids=new_whitelisted_channel_ids)
+
+        await self.save_permission_config_for_server(ctx.guild.id, permission, new_permission_config)
+
+        if new_permission_config.whitelisted_channel_ids:
+            # The permission is now restricted to the specified whitelisted channels.
+            update_text = UPDATE_RESTRICTED_FORMAT.format(permission.name) + \
+                          new_permission_config.get_channel_whitelist_display_text()
+            update_emoji = EMOJI_PERMISSION_RESTRICTED
+        else:
+            # The permission is now fully enabled in the server.
+            update_text = UPDATE_ENABLED_FORMAT.format(permission.name)
+            update_emoji = EMOJI_PERMISSION_ENABLED
+
+        await Permissions.announce_permission_updates(ctx, update_text, update_emoji)
+
+    async def get_permission_config_for_server(self, server_id: int, permission: Permission) -> PermissionConfig:
+        async with self.cache_lock:
+            if server_id not in self.cache:
+                self.cache[server_id] = {}
+
+            if permission.id not in self.cache[server_id]:
+                async with connect(self.db) as connection:
+                    query = 'SELECT * FROM permissions WHERE server_id=? AND permission_id=?'
+                    async with connection.execute(query, (server_id, permission.id)) as cursor:
+                        row = await cursor.fetchone()
+                        if row:
+                            permission_config = PermissionConfig.get_config(
+                                is_enabled=row[2], whitelisted_channel_ids=frozenset(loads(row[3])))
+                        else:
+                            permission_config = PermissionConfig.get_default_config_for_permission(permission)
+                self.cache[server_id][permission.id] = permission_config
+
+            return self.cache[server_id][permission.id]
+
+    async def save_permission_config_for_server(
+            self, server_id: int, permission: Permission, permission_config: PermissionConfig):
+        if permission_config == PermissionConfig.get_default_config_for_permission(permission):
+            # Setting to the default config is effectively the same as resetting the config, so favor the simpler path.
+            await self.reset_permission_config_for_server(server_id, permission)
+            return
+
+        async with self.cache_lock:
+            if server_id not in self.cache:
+                self.cache[server_id] = {}
+
+            self.cache[server_id][permission.id] = permission_config
+
+            async with connect(self.db) as connection:
+                is_enabled = permission_config.is_enabled
+                whitelisted_channel_ids = dumps(sorted(permission_config.whitelisted_channel_ids))
+                query = 'SELECT * FROM permissions WHERE server_id=? AND permission_id=?'
```

Comment:
Hmmm this may be more concisely represented with an [upsert](https://www.sqlite.org/lang_UPSERT.html) operator